### DISCUSSION TITLE
CPV control digit

### DISCUSSION THREAD

**Discussion by Defkon1** [eprocurementontology/eprocurementontology#589] [Q&A] [answered]
Hello everybody!

I'm looking for info on how to calculate the control digit of a CPV number to code a small validator...

Any hints?

**Comment by dancicr** [comment, marked as answer]
Please see the attached file.
[CPV_check digit calculation.docx](https://github.com/OP-TED/ePO/files/14742688/CPV_check.digit.calculation.docx)

**Reply by Defkon1** [reply]
Thanks!

**Comment by Defkon1** [comment]
Based on the document you attached, I'm understanding that the check digit is calculated using the sequence 7-1-3 on each level of the code hierarchy.

So, for example, the check digit for CPV Code 03300000-x should be:

Division (03) = 7 + 7 = 14 -> 4 + 7 = 11 -> 1
Group (3) = Division + 1 = 2 -> 2+1 = 3 -> 3+1 = 4

But, instead, that CPV code is listed as 03300000-2... 

What am I missing?

**Comment by dancicr** [comment]
Indeed, we have received the same input from other users of the documentation on the calculation of the check digit. Unfortunately, for such an error made by the colleagues who were responsible for the last revision of the CPV in 2008, there is nothing we can do at this time (the check digit is rarely used in public procurement). The CPV is anyway outdated and we are trying to replace it (with the help of EUROSTAT) with the CPA (link<https://ec.europa.eu/eurostat/statistics-explained/index.php?title=Glossary:Statistical_classification_of_products_by_activity_(CPA)>) as we do not have the expertise to revise/maintain classifications.

**Reply by Defkon1** [reply]
so, if I understand correctly, is actually impossible to write an algorithm to check formally valid CPV codes, isn't it?

EDIT/ based on my tests, the check digit error impacts 9291 codes out of 9454!

**Comment by dancicr** [comment]
Indeed, based on the feedback received ( I checked this also with my previous colleague responsible for CPV) the colleagues who implemented the last revision of the CPV in 2008 have made an error either in the documented formula to compute it or in its implementation. There is no information in our CPV related database to know what causes it or how it can be fixed.
